### PR TITLE
NOMERGEY Very early WIP stray vertex cleanup

### PR DESCRIPTION
**Pull request by helgridly**
I'd like to do a presentation-style review of this as I'd like feedback with the direction I'm going (I'm not super thrilled with any of this).

The general overview: Write a parallel implementation of `saveToVertex` called `deleteVertex` that handles deletes. This entails matching all cases where `saveToVertex` creates additional vertices and deleting them.

[Here](https://github.com/broadinstitute/rawls/pull/90/files#diff-e295baee74fdedfd2e5dbfd08f2e7265R249) is the entry point you should probably be starting at.

Some things to think about:

- Is any of this even necessary? I'm doing this whole mess of inspecting the class properties and recursively deleting, but the `removeMapVertex` method that recursively follows edges of type VertexSchema.Map and deletes them. If all containers are (or can be) represented like this then I'm very probably overengineering this solution.

- [This line](https://github.com/broadinstitute/rawls/pull/90/files#diff-e295baee74fdedfd2e5dbfd08f2e7265R291) really bugs me; see the FIXME comment. I think if you have an `Option[Seq[SomeDomainObject]]` and try to update its field from a `Some` to `None`, a recursive call to `removeMapVertex` won't be sufficient to remove the member vertices because they won't be of type `VertexSchema.Map`.

- The serialization code makes me really sad. I really think that our collection types should be recursively serialized; looking inside a `Seq` to determine how to serialize it horrifies me and I don't see why we should do it. I am really tempted to rewrite it all...

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 1]
```diff
@@ -205,6 +205,67 @@ trait GraphDAO {
     getSinglePipelineResult(workflowPipeline(workspaceContext, workflowId))
   }
 
+  def getProperties[T: TypeTag: ClassTag](obj: T): Iterable[(String, Option[Any])] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I pulled this out of `saveToVertex` as I need it for delete.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 2]
```diff
@@ -241,6 +289,9 @@ trait GraphDAO {
         // can't tell if this is a value or a map anymore so just remove both (there should only be one)
         vertex.removeProperty(key)
         vertex.getVertices(Direction.OUT, key).foreach(removeMapVertex)
+        //FIXME: ^^^ I think this is insufficient and doesn't cover the case where you overwrite a Some(foo:FooClass) with None, for a FooClass that requires sub-sub-vertices
+        //FIXME: See FIXME in getProperties - I think we can do this if we have the type too?
+        //Needs some shenanigans to figure out the type of an Option if it's None, see http://bit.ly/1htTZJk
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I think there's a similar issue to this going on in [line 299](https://github.com/broadinstitute/rawls/pull/90/files#diff-e295baee74fdedfd2e5dbfd08f2e7265R299); I don't think this isn't sufficient to clear up vertices left behind if we're overwriting a full list with an empty one.

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        txn.withGraph { graph =>
+          workspaceDAO.save(wsWithAttributeVals, txn)
+          workspaceDAO.delete(wsName, txn)
+          assertResult(0) {
+            graph.getVertices.toList.size
+          }
+        }
+      }
+    }
+
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala]: These tests are by no means exhaustive but they do at least now pass!

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import org.broadinstitute.dsde.rawls.graph.OrientDbTestFixture
+import org.broadinstitute.dsde.rawls.model._
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: hussein
+ * Date: 08/18/2015
+ * Time: 15:05
+ */
+class GraphDAODeleteSpec extends FreeSpec with Matchers with OrientDbTestFixture {
```
Comment (mattsooknah) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala]: I'd prefer adding tests to ```GraphWorkspaceDAOSpec```, etc instead of creating a new spec.

Comment (abaumann) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala]: since the tests are atomic i dont see a problem with smaller specs

Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAODeleteSpec.scala]: FWIW there are going to _very many_ delete tests, as there are a billion edge cases. This was just the start.

**Comment by mattsooknah** [comment]
Closing for now.
